### PR TITLE
[Merged by Bors] - Move dependencies to workspace

### PR DESCRIPTION
**Pull request by jxs**
## Issue Addressed

Synchronize dependencies and edition on the workspace `Cargo.toml`

## Proposed Changes

with https://github.com/rust-lang/cargo/issues/8415 merged it's now possible to synchronize details on the workspace `Cargo.toml` like the metadata and dependencies.
By only having dependencies that are shared between multiple crates aligned on the workspace `Cargo.toml` it's easier to not miss duplicate versions of the same dependency and therefore ease on the compile times.

## Additional Info
this PR also removes the no longer required direct dependency of the `serde_derive` crate.

should be reviewed after https://github.com/sigp/lighthouse/pull/4639 get's merged.
closes https://github.com/sigp/lighthouse/issues/4651

### CODE REVIEW

File: Cargo.toml
Comment (jimmygchen) [inline Cargo.toml:221]: nitpick:

```suggestion
validator_client = { path = "validator_client" }
```

**Review by jimmygchen — APPROVED** [review]
> this PR also removes the no longer required serde_derive crate.
What do you mean by this? `serde_derive` is still used and looks like this is one of the few crates that wasn't updated to workspace dependency. I'm curious about the reason behind this?

This is really nice! I'm not super familiar with cargo dependency management, but using workspace dependencies looks really neat, and I noticed a quite few duplicate dependencies have been removed which gives us more consistency in dependency versions and makes it a lot easier to maintain. 

I don't really have any other comments, the looks really good and tests are all passing so :clap:

File: Cargo.toml
Comment (jxs) [reply Cargo.toml:221]: thanks Jimmy!

**Comment by jxs** [comment]
> > this PR also removes the no longer required serde_derive crate.
> > What do you mean by this? `serde_derive` is still used and looks like this is one of the few crates that wasn't updated to workspace dependency. I'm curious about the reason behind this?
> 
> This is really nice! I'm not super familiar with cargo dependency management, but using workspace dependencies looks really neat, and I noticed a quite few duplicate dependencies have been removed which gives us more consistency in dependency versions and makes it a lot easier to maintain.
> 
> I don't really have any other comments, the looks really good and tests are all passing so 👏

yeah sorry if it was misleading, what I meant regarding `serde-derive` is that we wouldn't be required to depend directly on it anymore. Updated the description to address this

**Review by paulhauner — COMMENTED** [review]
Looks good to me, big effort!

Some of the crates in this repo are [published to crates.io](https://crates.io/teams/github:sigp:crates-io), I wonder if they're still publishable with `workspace = true` rather than the full dependency definition? I'm happy to approve and merge if they can still be published!

**Comment by michaelsproul** [comment]
> Some of the crates in this repo are [published to crates.io](https://crates.io/teams/github:sigp:crates-io)

Are they? I think we already moved all of them out? All of `enr`, `discv5`, `ssz_types`, `ethereum_ssz`, `tree_hash`, `ethereum_hashing`, etc are in external repos:

- https://github.com/sigp/enr
- https://github.com/sigp/discv5
- https://github.com/sigp/ssz_types
- https://github.com/sigp/ethereum_ssz
- https://github.com/sigp/tree_hash
- https://github.com/sigp/ethereum_hashing

File: Cargo.toml
Comment (michaelsproul) [inline Cargo.toml]: We can (and should) turn off the `blocking` feature now that https://github.com/sigp/lighthouse/pull/4753 is merged

Comment (jxs) [reply Cargo.toml]: yeah, thanks Michael, I was waiting for https://github.com/ralexstokes/ethereum-consensus/pull/261 to then switch `native-tls-vendored` for just `rustls-tls` but we can do that on a subsequent PR. Updated!

**Comment by jxs** [comment]
> Looks good to me, big effort!
> 
> Some of the crates in this repo are [published to crates.io](https://crates.io/teams/github:sigp:crates-io), I wonder if they're still publishable with `workspace = true` rather than the full dependency definition? I'm happy to approve and merge if they can still be published!

Thanks Paul, yeah afaik it's possible, [`rust-libp2p`](https://github.com/libp2p/rust-libp2p/) has the same architecture and publishes the subcrates.

File: Cargo.toml
Comment (jxs) [reply Cargo.toml]: https://github.com/sigp/lighthouse/actions/runs/6268152351/job/17022546605?pr=4650#step:8:54 seems to fail without `blocking`, reverted. I can address it in a subsequent PR so as to not block this one

Comment (michaelsproul) [reply Cargo.toml]: oh yeah no worries!

File: Cargo.toml
Comment (michaelsproul) [inline Cargo.toml]:
```suggestion
reqwest = { version = "0.11", default-features = false, features = ["blocking", "json", "stream", "rustls-tls"] }
```

Comment (jimmygchen) [reply Cargo.toml]: nice catch, I missed this one 😱

Comment (michaelsproul) [reply Cargo.toml]: it just got added in the last few hours

**Review by michaelsproul — APPROVED** [review]
Let's merge!!

**Comment by michaelsproul** [comment]
Ah, there's a tokio dep issue introduced by 893b3eb6639e67ff7657d0b95a1c2dec6549f6a2. I'll fix that, test locally, and push

**Review by jimmygchen — APPROVED** [review]

**Comment by michaelsproul** [comment]
Ah this is failing because of the web3signer tests and the openssl upgrade. I'll remove it from the current batch and see if I can get it passing by holding back openssl

bors r-

**Comment by michaelsproul** [comment]
Ok I think I got it to work without holding any deps. I just regenerated the web3signer certs using `./generate.sh`. This seems to be enough to make the test pass on my machine. Presumably the new version of openssl defaulted to a more modern non-deprecated algorithm.

**Comment by paulhauner** [comment]
bors retry

**Comment by jxs** [comment]
Thanks for the help Michael :)

File: beacon_node/execution_layer/Cargo.toml
Comment (ralexstokes) [inline beacon_node/execution_layer/Cargo.toml:48]: fwiw I've re-exported the `ssz_rs` `prelude` to `ethereum_consensus::ssz::prelude` if you want to drop this dep

**Pull request closed without merging**
